nexus/reconfigurator/preparation/src/lib.rs

```diff
@@ -72,3 +89,147 @@ pub fn policy_from_db(
+            .await
+            .context("listing services IP pool ranges")?
+    };
+
+    let policy = policy_from_db(
+        &sled_rows,
+        &zpool_rows,
+        &ip_pool_range_rows,
+        NEXUS_REDUNDANCY,
+    )
+    .context("assembling policy")?;
+
+    let collection_ids = datastore
+        .inventory_collections()
+        .await
+        .context("listing collections")?;
+    let collections = futures::stream::iter(collection_ids)
+        .filter_map(|id| async move {
+            let read = datastore
+                .inventory_collection_read(opctx, id)
+                .await
+                .with_context(|| format!("reading collection {}", id));
+            // It's not necessarily a problem if we failed to read a collection.
+            // They can be removed since we fetched the list.
+            read.ok()
+        })
+        .collect::<Vec<Collection>>()
+        .await;
+
+    let mut blueprint_ids = Vec::new();
+    let mut paginator = Paginator::new(SQL_BATCH_SIZE);
+    while let Some(p) = paginator.next() {
+        let batch = datastore
+            .blueprints_list(opctx, &p.current_pagparams())
+            .await
+            .context("listing blueprints")?;
+        paginator =
+            p.found_batch(&blueprint_ids, &|b: &BlueprintMetadata| b.id);
+        blueprint_ids.extend(batch.into_iter());
+    }
+
+    let blueprints = futures::stream::iter(blueprint_ids)
+        .filter_map(|bpm| async move {
+            let blueprint_id = bpm.id;
+            let read = datastore
+                .blueprint_read(
+                    opctx,
+                    &nexus_db_queries::authz::Blueprint::new(
+                        nexus_db_queries::authz::FLEET,
+                        blueprint_id,
+                        LookupType::ById(blueprint_id),
+                    ),
+                )
+                .await
+                .with_context(|| format!("reading blueprint {}", blueprint_id));
+            // It's not necessarily a problem if we failed to read a blueprint.
+            // They can be removed since we fetched the list.
+            read.ok()
+        })
+        .collect::<Vec<Blueprint>>()
+        .await;
+
+    // It's also useful to include information about any DNS generations
+    // mentioned in any blueprints.
+    let blueprints_list = &blueprints;
+    let fetch_dns_group = |dns_group: DnsGroup| async move {
+        let latest_version = datastore
+            .dns_group_latest_version(&opctx, dns_group)
+            .await
+            .with_context(|| {
+                format!("reading latest {:?} version", dns_group)
+            })?;
+        let dns_generations_needed: BTreeSet<_> = blueprints_list
+            .iter()
+            .map(|blueprint| match dns_group {
+                DnsGroup::Internal => blueprint.internal_dns_version,
+                DnsGroup::External => blueprint.external_dns_version,
+            })
```

Comment:
Should this live as a method on `Blueprint`?

Reply:
Conceivably, though that's not trivial today because `DnsGroup` is in `nexus-db-model`, which depends on `nexus-types` and not the other way around.

Reply:
Ahh yeah. Yay. Maybe if the need arises we can add a separate `DnsGroup` to nexus-types in the future.